storage/xtradb/dict/dict0dict.cc

```diff
@@ -941,12 +952,20 @@ dict_index_contains_col_or_prefix(
 		return(TRUE);
 	}
 
-	col = dict_table_get_nth_col(index->table, n);
+	if (dict_index_is_unique_hash(index)) {
+		n_fields = dict_index_get_n_user_fields(index);
+	} else {
+		n_fields = dict_index_get_n_fields(index);
+	}
 
-	n_fields = dict_index_get_n_fields(index);
+	col = dict_table_get_nth_col(index->table, n);
 
 	for (pos = 0; pos < n_fields; pos++) {
-		field = dict_index_get_nth_field(index, pos);
+		if (dict_index_is_unique_hash(index)) {
```

Comment:
You could have a inlined function to to this as this same repeats several times.